tinygo fails to build at Arch Linux clean environment

I am building this project in a Arch Linux clean environment (kinda chroot with sanitized environment) and the build fails:

```
[158/230] Building CXX object third_party/llvm-project/CMakeFiles/llvm_dwarf.dir/DataExtractor.cpp.o
[159/230] Building CXX object third_party/llvm-project/CMakeFiles/llvm_dwarf.dir/DJB.cpp.o
[160/230] Building CXX object src/wasm/CMakeFiles/wasm.dir/wasm-type.cpp.o
[161/230] Building CXX object src/passes/CMakeFiles/passes.dir/SimplifyLocals.cpp.o
FAILED: src/passes/CMakeFiles/passes.dir/SimplifyLocals.cpp.o 
/usr/bin/c++  -I/build/tinygo/src/tinygo/lib/binaryen/src -I/build/tinygo/src/tinygo/lib/binaryen/third_party/llvm-project/include -I/build/tinygo/src/tinygo/lib/binaryen -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=3 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection         -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -Wp,-D_GLIBCXX_ASSERTIONS -g -ffile-prefix-map=/build/tinygo/src=/usr/src/debug/tinygo -DBUILD_LLVM_DWARF -Wall -Werror -Wextra -Wno-unused-parameter -fno-omit-frame-pointer -fno-rtti -Wno-implicit-int-float-conversion -Wno-unknown-warning-option -Wswitch -Wimplicit-fallthrough -Wnon-virtual-dtor -fPIC -fdiagnostics-color=always -O3 -DNDEBUG -UNDEBUG -std=c++17 -MD -MT src/passes/CMakeFiles/passes.dir/SimplifyLocals.cpp.o -MF src/passes/CMakeFiles/passes.dir/SimplifyLocals.cpp.o.d -o src/passes/CMakeFiles/passes.dir/SimplifyLocals.cpp.o -c /build/tinygo/src/tinygo/lib/binaryen/src/passes/SimplifyLocals.cpp
In file included from /build/tinygo/src/tinygo/lib/binaryen/src/parsing.h:29,
                 from /build/tinygo/src/tinygo/lib/binaryen/src/wasm-builder.h:21,
                 from /build/tinygo/src/tinygo/lib/binaryen/src/ir/literal-utils.h:20,
                 from /build/tinygo/src/tinygo/lib/binaryen/src/ir/bits.h:21,
                 from /build/tinygo/src/tinygo/lib/binaryen/src/ir/properties.h:20,
                 from /build/tinygo/src/tinygo/lib/binaryen/src/ir/iteration.h:20,
                 from /build/tinygo/src/tinygo/lib/binaryen/src/ir/branch-utils.h:20,
                 from /build/tinygo/src/tinygo/lib/binaryen/src/passes/SimplifyLocals.cpp:50:
In constructor ‘wasm::Walker<SubType, VisitorType>::Task::Task(wasm::Walker<SubType, VisitorType>::TaskFunc, wasm::Expression**) [with SubType = wasm::LocalGetCounter; VisitorType = wasm::Visitor<wasm::LocalGetCounter, void>]’,
    inlined from ‘void wasm::SmallVector<T, N>::emplace_back(ArgTypes&& ...) [with ArgTypes = {void (*&)(wasm::LocalGetCounter*, wasm::Expression**), wasm::Expression**&}; T = wasm::Walker<wasm::LocalGetCounter, wasm::Visitor<wasm::LocalGetCounter, void> >::Task; long unsigned int N = 10]’ at /build/tinygo/src/tinygo/lib/binaryen/src/support/small_vector.h:73:7,
    inlined from ‘void wasm::Walker<SubType, VisitorType>::pushTask(TaskFunc, wasm::Expression**) [with SubType = wasm::LocalGetCounter; VisitorType = wasm::Visitor<wasm::LocalGetCounter, void>]’ at /build/tinygo/src/tinygo/lib/binaryen/src/wasm-traversal.h:314:23,
    inlined from ‘void wasm::Walker<SubType, VisitorType>::walk(wasm::Expression*&) [with SubType = wasm::LocalGetCounter; VisitorType = wasm::Visitor<wasm::LocalGetCounter, void>]’ at /build/tinygo/src/tinygo/lib/binaryen/src/wasm-traversal.h:329:13,
    inlined from ‘void wasm::LocalGetCounter::analyze(wasm::Function*, wasm::Expression*)’ at /build/tinygo/src/tinygo/lib/binaryen/src/ir/local-utils.h:37:9:
/build/tinygo/src/tinygo/lib/binaryen/src/wasm-traversal.h:309:59: error: storing the address of local variable ‘ast’ in ‘*(std::array<wasm::Walker<wasm::LocalGetCounter, wasm::Visitor<wasm::LocalGetCounter, void> >::Task, 10>*)((char*)&_2->stack + 8).std::array<wasm::Walker<wasm::LocalGetCounter, wasm::Visitor<wasm::LocalGetCounter, void> >::Task, 10>::_M_elems[<unknown>].wasm::Walker<wasm::LocalGetCounter, wasm::Visitor<wasm::LocalGetCounter, void> >::Task::currp’ [-Werror=dangling-pointer=]
  309 |     Task(TaskFunc func, Expression** currp) : func(func), currp(currp) {}
      |                                                           ^~~~~~~~~~~~
In file included from /build/tinygo/src/tinygo/lib/binaryen/src/passes/SimplifyLocals.cpp:54:
/build/tinygo/src/tinygo/lib/binaryen/src/ir/local-utils.h: In member function ‘void wasm::LocalGetCounter::analyze(wasm::Function*, wasm::Expression*)’:
/build/tinygo/src/tinygo/lib/binaryen/src/ir/local-utils.h:34:44: note: ‘ast’ declared here
   34 |   void analyze(Function* func, Expression* ast) {
      |                                ~~~~~~~~~~~~^~~
/build/tinygo/src/tinygo/lib/binaryen/src/ir/local-utils.h:34:47: note: ‘this’ declared here
   34 |   void analyze(Function* func, Expression* ast) {
      |                                               ^
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-implicit-int-float-conversion’ may have been intended to silence earlier diagnostics
cc1plus: all warnings being treated as errors
[162/230] Building CXX object third_party/llvm-project/CMakeFiles/llvm_dwarf.dir/Dwarf.cpp.o
[163/230] Building CXX object third_party/llvm-project/CMakeFiles/llvm_dwarf.dir/DWARFAbbreviationDeclaration.cpp.o
[164/230] Building CXX object src/analysis/CMakeFiles/analysis.dir/cfg.cpp.o
```


The `binaryen` submodule fails to compile `SimplifyLocals.cpp` because compile complains about unknown options.

The error comes from the fact that the default compiler was picked up, and in the chroot environment that compiler is `/usr/bin/c++` that is currently GCC 14.1.1.

`binaryen` on the other hand does not work with GCC. It looks like it expects to be compiled with `clang`. In this case `GNUmakefile` should reflect it. Setting proper `CMAKE_CXX_COMPILER` cmake variable should help here I think.

Hi @aykevl thank you for your reply.

I built `binaryen` separately with your instructions and it compiles fine at Arch Linux x86_64.

So the problem is most likely comes from the way the build environment set up. Among other things it sets compilation flags. I compiled `binaryen` with those flags and now it fails:

```
$ export CXXFLAGS="-O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=3 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection         -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -Wp,-D_GLIBCXX_ASSERTIONS -g"
$ cmake -G Ninja . -DBUILD_TESTS=OFF
-- The C compiler identification is GNU 14.1.1
-- The CXX compiler identification is GNU 14.1.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- No build type selected, default to Release
-- Building with -DBUILD_LLVM_DWARF
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Building with -Wall
-- Building with -Werror
-- Building with -Wextra
-- Building with -Wno-unused-parameter
-- Building with -fno-omit-frame-pointer
-- Building with -fno-rtti
-- Building with -Wno-implicit-int-float-conversion
-- Building with -Wno-unknown-warning-option
-- Building with -Wswitch
-- Building with -Wimplicit-fallthrough
-- Building with -Wnon-virtual-dtor
-- Building with -fPIC
-- Building with -UNDEBUG
-- Building with -fdiagnostics-color=always
-- Building libbinaryen as shared library.
-- Configuring done (0.5s)
-- Generating done (0.0s)
-- Build files have been written to: lib/binaryen
$ ninja bin/wasm-opt -j9
...
[161/230] Building CXX object src/passes/CMakeFiles/passes.dir/SimplifyLocals.cpp.o
FAILED: src/passes/CMakeFiles/passes.dir/SimplifyLocals.cpp.o 
/usr/bin/c++  -I/home/anatol/sources/tinygo/tinygo/lib/binaryen/src -I/home/anatol/sources/tinygo/tinygo/lib/binaryen/third_party/llvm-project/include -I/home/anatol/sources/tinygo/tinygo/lib/binaryen -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=3 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection         -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -Wp,-D_GLIBCXX_ASSERTIONS -g -DBUILD_LLVM_DWARF -Wall -Werror -Wextra -Wno-unused-parameter -fno-omit-frame-pointer -fno-rtti -Wno-implicit-int-float-conversion -Wno-unknown-warning-option -Wswitch -Wimplicit-fallthrough -Wnon-virtual-dtor -fPIC -fdiagnostics-color=always -O3 -DNDEBUG -UNDEBUG -std=c++17 -MD -MT src/passes/CMakeFiles/passes.dir/SimplifyLocals.cpp.o -MF src/passes/CMakeFiles/passes.dir/SimplifyLocals.cpp.o.d -o src/passes/CMakeFiles/passes.dir/SimplifyLocals.cpp.o -c /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/passes/SimplifyLocals.cpp
In file included from /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/parsing.h:29,
                 from /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/wasm-builder.h:21,
                 from /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/ir/literal-utils.h:20,
                 from /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/ir/bits.h:21,
                 from /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/ir/properties.h:20,
                 from /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/ir/iteration.h:20,
                 from /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/ir/branch-utils.h:20,
                 from /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/passes/SimplifyLocals.cpp:50:
In constructor ‘wasm::Walker<SubType, VisitorType>::Task::Task(wasm::Walker<SubType, VisitorType>::TaskFunc, wasm::Expression**) [with SubType = wasm::LocalGetCounter; VisitorType = wasm::Visitor<wasm::LocalGetCounter, void>]’,
    inlined from ‘void wasm::SmallVector<T, N>::emplace_back(ArgTypes&& ...) [with ArgTypes = {void (*&)(wasm::LocalGetCounter*, wasm::Expression**), wasm::Expression**&}; T = wasm::Walker<wasm::LocalGetCounter, wasm::Visitor<wasm::LocalGetCounter, void> >::Task; long unsigned int N = 10]’ at /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/support/small_vector.h:73:7,
    inlined from ‘void wasm::Walker<SubType, VisitorType>::pushTask(TaskFunc, wasm::Expression**) [with SubType = wasm::LocalGetCounter; VisitorType = wasm::Visitor<wasm::LocalGetCounter, void>]’ at /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/wasm-traversal.h:314:23,
    inlined from ‘void wasm::Walker<SubType, VisitorType>::walk(wasm::Expression*&) [with SubType = wasm::LocalGetCounter; VisitorType = wasm::Visitor<wasm::LocalGetCounter, void>]’ at /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/wasm-traversal.h:329:13,
    inlined from ‘void wasm::LocalGetCounter::analyze(wasm::Function*, wasm::Expression*)’ at /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/ir/local-utils.h:37:9:
/home/anatol/sources/tinygo/tinygo/lib/binaryen/src/wasm-traversal.h:309:59: error: storing the address of local variable ‘ast’ in ‘*(std::array<wasm::Walker<wasm::LocalGetCounter, wasm::Visitor<wasm::LocalGetCounter, void> >::Task, 10>*)((char*)&_2->stack + 8).std::array<wasm::Walker<wasm::LocalGetCounter, wasm::Visitor<wasm::LocalGetCounter, void> >::Task, 10>::_M_elems[<unknown>].wasm::Walker<wasm::LocalGetCounter, wasm::Visitor<wasm::LocalGetCounter, void> >::Task::currp’ [-Werror=dangling-pointer=]
  309 |     Task(TaskFunc func, Expression** currp) : func(func), currp(currp) {}
      |                                                           ^~~~~~~~~~~~
In file included from /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/passes/SimplifyLocals.cpp:54:
/home/anatol/sources/tinygo/tinygo/lib/binaryen/src/ir/local-utils.h: In member function ‘void wasm::LocalGetCounter::analyze(wasm::Function*, wasm::Expression*)’:
/home/anatol/sources/tinygo/tinygo/lib/binaryen/src/ir/local-utils.h:34:44: note: ‘ast’ declared here
   34 |   void analyze(Function* func, Expression* ast) {
      |                                ~~~~~~~~~~~~^~~
/home/anatol/sources/tinygo/tinygo/lib/binaryen/src/ir/local-utils.h:34:47: note: ‘this’ declared here
   34 |   void analyze(Function* func, Expression* ast) {
      |                                               ^
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-implicit-int-float-conversion’ may have been intended to silence earlier diagnostics
cc1plus: all warnings being treated as errors
[169/230] Building CXX object src/wasm/CMakeFiles/wasm.dir/wat-parser.cpp.o
ninja: build stopped: subcommand failed.
```

Actually it seems that the real error is
```
/home/anatol/sources/tinygo/tinygo/lib/binaryen/src/wasm-traversal.h:309:59: error: storing the address of local variable ‘ast’ in ‘*(std::array<wasm::Walker<wasm::LocalGetCounter, wasm::Visitor<wasm::LocalGetCounter, void> >::Task, 10>*)((char*)&_2->stack + 8).std::array<wasm::Walker<wasm::LocalGetCounter, wasm::Visitor<wasm::LocalGetCounter, void> >::Task, 10>::_M_elems[<unknown>].wasm::Walker<wasm::LocalGetCounter, wasm::Visitor<wasm::LocalGetCounter, void> >::Task::currp’ [-Werror=dangling-pointer=]
  309 |     Task(TaskFunc func, Expression** currp) : func(func), currp(currp) {}
      |                                                           ^~~~~~~~~~~~
In file included from /home/anatol/sources/tinygo/tinygo/lib/binaryen/src/passes/SimplifyLocals.cpp:54:
/home/anatol/sources/tinygo/tinygo/lib/binaryen/src/ir/local-utils.h: In member function ‘void wasm::LocalGetCounter::analyze(wasm::Function*, wasm::Expression*)’:
/home/anatol/sources/tinygo/tinygo/lib/binaryen/src/ir/local-utils.h:34:44: note: ‘ast’ declared here
```

that is probably cause by one of the security fortification flags.

Does the compiler error look legit?

--- a/GNUmakefile
+++ b/GNUmakefile
@@ -257,7 +257,7 @@ ifneq ($(USE_SYSTEM_BINARYEN),1)
 binaryen: build/wasm-opt$(EXE)
 build/wasm-opt$(EXE):
 	mkdir -p build
-	cd lib/binaryen && cmake -G Ninja . -DBUILD_STATIC_LIB=ON -DBUILD_TESTS=OFF $(BINARYEN_OPTION) && ninja bin/wasm-opt$(EXE)
+	cd lib/binaryen && cmake -G Ninja . -DBUILD_STATIC_LIB=ON -DBUILD_TESTS=OFF -DENABLE_WERROR=OFF $(BINARYEN_OPTION) && ninja bin/wasm-opt$(EXE)
 	cp lib/binaryen/bin/wasm-opt$(EXE) build/wasm-opt$(EXE)
 endif
 

